### PR TITLE
feature/20692-treemap-small-nodes-grouping

### LINKED ISSUE #20692: Group small leaf nodes in Treemap series.
Optimize treemap rendering by grouping smaller leaf nodes below a certain threshold into a single point with ellipses. This should work for leaves sharing the same parent, to manage high node counts.

User's original description:

>  Do not render all points for optimization
If no. of points high then we can should them like this (just an example) -
image
Ideating about the feature -
It can be applicable for all leaf nodes having the same parent
If no. of leaf nodes (under a parent) are high, then node having lesser area can be combined together based on a threshold e.g. maxNodesToShow, rest nodes can be shown like in above image with ellipses.

# Description of the feature

# Library related to the feature

# Proof of Concept/Live example for the feature
image

**You can vote for this feature by adding a thumbs-up reaction to this post.**

### PR DESCRIPTION
**Pull request by hubertkozik**
Added new treemap feature, the grouping of small leaf nodes, configured by the [treemap.cluster](https://api.highcharts.com/highcharts/series.treemap.cluster) option.

---

Closes #20692.

- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206668318846607

### CODE REVIEW

**Review by KacperMadej — CHANGES_REQUESTED** [review]
- [x] Click on the grouped after drilling down to the top left category in highcharts/plotoptions/treemap-grouping-advanced is causing an error in the console.
- [ ] Both bottom right groups only partially fill the available space (like 50%) after a drilldown.
image
image
- [x] drawing `dataLabels` in the advanced demo takes over 8 seconds
image

File: ts/Series/Treemap/TreemapSeries.ts [hunk 1]
```diff
@@ -427,6 +427,158 @@ class TreemapSeries extends ScatterSeries {
...
+                        )
+                    ) ||
+                    (
+                        !compareHeight &&
+                        (
+                            a < pixelWidth ||
+                            b < pixelWidth ||
+                            area < thresholdArea
+                        )
+                    )
```
[the remark this answers is not in the document]
Comment (hubertkozik) [reply ts/Series/Treemap/TreemapSeries.ts:503]: Right, I didn't simplify it! Fixed.

File: ts/Series/Treemap/TreemapNode.ts [hunk 2]
```diff
@@ -37,6 +37,7 @@ class TreemapNode {
     childrenTotal = 0;
     ignore?: boolean;
     isLeaf?: boolean;
+    isGroup: boolean = false;
```
Comment (bre1470) [inline ts/Series/Treemap/TreemapNode.ts:40]: You can make it optional and skip the initial assignment.

File: ts/Series/Treemap/TreemapSeriesDefaults.ts [hunk 3]
```diff
@@ -209,8 +209,25 @@ const TreemapSeriesDefaults: TreemapSeriesOptions = {
...
+        /**
+         * The HTML of the grouped nodes point's in the tooltip. Works only for
+         * Treemap series grouping and analogously to
+         * [pointFormat](#tooltip.pointFormat).
+         *
+         * The grouped nodes point tooltip can be also formatted using
+         * `tooltip.formatter` callback function and `point.isGroupNode` flag.
+         *
+         * @type      {string}
+         * @default   + {point.groupedPointsAmount} more...
```
Comment (bre1470) [inline ts/Series/Treemap/TreemapSeriesDefaults.ts:226]: C&P problem with `+` and `more...`. This should probably be a string.

File: ts/Series/Treemap/TreemapSeriesOptions.d.ts [hunk 4]
```diff
@@ -106,6 +106,55 @@ export interface TreemapSeriesLevelColorVariationOptions {
...
+
+    /**
+     * The pixel threshold width of area, which is used in Treemap grouping.
+     *
+     * @type {number}
+     *
+     * @since next
+     *
+     * @product highcharts
+     *
```
Comment (bre1470) [inline ts/Series/Treemap/TreemapSeriesOptions.d.ts:130]: Unnecessary spacing as in others as well. Keep it compact for one-liners.

File: ts/Series/Treemap/TreemapSeriesOptions.d.ts [hunk 5]
```diff
@@ -591,7 +660,20 @@ export interface TreemapSeriesOptions extends ScatterSeriesOptions {
...
+    /**
+     * The HTML of the grouped nodes point's in the tooltip. Works only for
+     * Treemap series grouping and analogously to
+     * [pointFormat](#tooltip.pointFormat).
+     *
+     * The grouped nodes point tooltip can be also formatted using
+     * `tooltip.formatter` callback function and `point.isGroupNode` flag.
+     *
+     * @type      {string}
+     * @default   + {point.groupedPointsAmount} more...
```
Comment (bre1470) [inline ts/Series/Treemap/TreemapSeriesOptions.d.ts:674]: The plus-more-thing ones more.

File: ts/Series/Treemap/TreemapSeriesDefaults.ts [hunk 6]
```diff
@@ -542,7 +559,69 @@ const TreemapSeriesDefaults: TreemapSeriesOptions = {
...
+         */
+
+        /**
+         * The name of the point of grouped nodes shown in the tooltip,
+         * dataLabels, etc.
+         *
+         * @type {string}
+         *
+         * @since next
+         *
```
Comment (bre1470) [inline ts/Series/Treemap/TreemapSeriesDefaults.ts:620]: Unnecessary spacing

**Review by bre1470 — CHANGES_REQUESTED** [review]

**Review by KacperMadej — APPROVED** [review]
Looks good to me.
Isn't the format "+ X more..." a bit misleading? I would expect to the group to contain **over** X and not exactly X.

**Review by MarkusBarstad — APPROVED** [review]
Very well done! 🚀

**Review by bre1470 — APPROVED** [review]
In the grouping demo the rest tooltip says "New tooltip content", which is a bit strange.
And I agree with what Kacper wrote.

Otherwise it looks good. 👍

File: samples/highcharts/plotoptions/treemap-grouping-advanced/demo.js [hunk 7]
```diff
@@ -0,0 +1,118 @@
+(async () => {
+
+    const dataJson = await fetch(
+        'https://cdn.jsdelivr.net/gh/highcharts/highcharts@8a1f8f0eb4/samples/data/africa-export-2021.json'
+    ).then(response => response.json());
+
+    const lightenDarkenColor = (col, amt) =>
```
Comment (TorsteinHonsi) [inline samples/highcharts/plotoptions/treemap-grouping-advanced/demo.js:7]: We have the [Highcharts.Color.brighten](https://api.highcharts.com/class-reference/Highcharts.Color#brighten) function that we use in some demos that looks a bit friendlier...

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Thanks, great work! I added some comments inline. 

- [x] Also, I think this feature deserves a mention in the `docs/chart-and-series-types/treemap.md` article.

- [x] I addition, did you consider adding a feature to click the grouped item to expand it? I am not sure how that would work, if it should expand to the parent level or only the grouped items. But this would be equivalent to clicking a [marker cluster in scatter charts](https://jsfiddle.net/gh/get/library/pure/highcharts/highcharts/tree/master/samples/highcharts/marker-clusters/basic/), where the cluster dissolves into the individual points.

File: ts/Series/Treemap/TreemapSeriesDefaults.ts [hunk 8]
```diff
@@ -542,7 +559,67 @@ const TreemapSeriesDefaults: TreemapSeriesOptions = {
...
+
+        /**
+         * The name of the point of grouped nodes shown in the tooltip,
+         * dataLabels, etc.
+         *
+         * @type {string}
+         * @since next
+         * @product highcharts
+         */
+        name: 'Grouped nodes'
```
[the remark this answers is not in the document]
Comment (hubertkozik) [reply ts/Series/Treemap/TreemapSeriesDefaults.ts:621]: Agree, removed default name and added info in API description.

**Review by TorsteinHonsi — COMMENTED** [review]
Thanks so far! I've been fiddling a bit with the traversing and have some comments/questions:

1. Consider highcharts/plotoptions/treemap-grouping-advanced. If you click the "All machinery" group, the nested nodes expand, but they seem to do so without animation. As far as I can see, this is contrary to the `treemap-large` demo, where animation occurs. 
2. Next, click the +168 tile. Also no animation.
3. The breadcrumbs now say +112. But I clicked +168...
4. Click Machinery again in the breadcrumbs. Now the +168 tile takes up som 30-40% of the plot area, unlike in step 1 where it was much smaller. 
5. Click +168 again. Now the breadcrumbs say +153.

All in all we may have to work a bit on animation. For comparison, see the highcharts/marker-clusters/basic demo, where all animations flow nicely, all points originating in the cluster. But this may not be unique to this PR or the new feature, but maybe a treemap issue by itself.

**Comment by hubertkozik** [comment]
@TorsteinHonsi I fixed the animation problem, turns out it was the `animationLimit` for this series in this demo.

Also, I fixed the problem of grouped node wrong sizing - it was much more complicated and I figured out, that the previous implementation of grouping was done in the wrong place, I refactored it one more time.

Sadly, now after traversing into grouped nodes their children sometimes look strange, but it's a correct layout behaviour - not sure if we should improve that. For example, after traversing into a grouped node in "Machinery" group there are some empty points, due to layout change/re-calculation, but only good points are shown on the chart.

**Comment by TorsteinHonsi** [comment]
> To be honest I am not sure if we should provide the possibility to traverse into grouped points, due to algorithm calculations it gets very complicated.

@hubertkozik Could you make a team decision on this in the core group? The client probably has a preference. If we can't decide clearly for one solution or the other, we could introduce an option.

File: ts/Series/Treemap/TreemapSeriesDefaults.ts [hunk 9]
```diff
@@ -542,7 +561,91 @@ const TreemapSeriesDefaults: TreemapSeriesOptions = {
...
+     *         Simple demo of Treemap grouping
+     * @sample {highcharts} highcharts/plotoptions/treemap-grouping-multiple-parents
+     *         Treemap grouping with multiple parents
+     * @sample {highcharts} highcharts/plotoptions/treemap-grouping-advanced
+     *         Advanced demo of Treemap grouping
+     *
+     * @since next
+     *
+     * @product highcharts
+     * @apioption plotOptions.treemap.groupAreaThreshold
```
Comment (bre1470) [inline ts/Series/Treemap/TreemapSeriesDefaults.ts:583]: You could skip `@apioption` as it is part of an `@optionparent`-tree.

**Review by bre1470 — APPROVED** [review]
Great job! Looks good. 👍

**Review by KacperMadej — APPROVED** [review]

File: samples/highcharts/plotoptions/treemap-grouping-advanced/demo.js [hunk 10]
```diff
@@ -0,0 +1,117 @@
+(async () => {
+
+    const dataJson = await fetch(
```
Comment (TorsteinHonsi) [inline samples/highcharts/plotoptions/treemap-grouping-advanced/demo.js:3]: Just `data`, it is no longer JSON after the JSON string has been parsed.
```suggestion
    const dataJson = await fetch(
```

File: ts/Series/Treemap/TreemapSeriesDefaults.ts [hunk 11]
```diff
@@ -210,7 +212,21 @@ const TreemapSeriesDefaults: TreemapSeriesOptions = {
...
+         * [pointFormat](#tooltip.pointFormat).
+         *
+         * The grouped nodes point tooltip can be also formatted using
+         * `tooltip.formatter` callback function and `point.isGroupNode` flag.
+         *
+         * @type      {string}
+         * @default   '+ {point.groupedPointsAmount} more...'
+         * @apioption tooltip.clusterFormat
+         */
+        clusterFormat: '<span>+ {point.groupedPointsAmount} more...' +
```
Comment (TorsteinHonsi) [inline ts/Series/Treemap/TreemapSeriesDefaults.ts:228]: Why the span?

File: ts/Series/Treemap/TreemapSeriesDefaults.ts [hunk 12]
```diff
@@ -542,7 +558,120 @@ const TreemapSeriesDefaults: TreemapSeriesOptions = {
...
      */
-    traverseToLeaf: false
+    traverseToLeaf: false,
+
+    /**
+     * An option to optimize treemap series rendering by grouping smaller leaf
+     * nodes below a certain square area threshold in pixels. If the square area
+     * of a point becomes smaller than the specified threshold, determined by
+     * the `pixelWidth` and/or `pixelHeight` options, then this point is moved
+     * into group point per series.
```
Comment (TorsteinHonsi) [inline ts/Series/Treemap/TreemapSeriesDefaults.ts:568]: Did you mean into _one_ group point?
```suggestion
     * into one group point per series.
```

**Review by TorsteinHonsi — CHANGES_REQUESTED** [review]
Great work! Just some minor details/questions.

File: ts/Series/Treemap/TreemapSeriesDefaults.ts (hunk 11 above)
Comment (hubertkozik) [reply ts/Series/Treemap/TreemapSeriesDefaults.ts:228]: Right, span not needed.

**Review by TorsteinHonsi — APPROVED** [review]

**Pull request merged**